### DISCUSSION TITLE
Best way to get a token with login/pass and reuse in all requests

### DISCUSSION THREAD

**Discussion by labomatik** [saloonphp/saloon#172] [Q&A] [answered]
Hello,

I would like to know the best way to use a Bearer token for all requests that is obtained with a login/pass:
1: get Bearer by sending a login/pass to API/token endpoint,
2: set the Bearer to all requests

I've created a new Authenticator but i'm not sure i should mix requests inside that class...

What is the best way to proceed?

Thanks

**Comment by Sammyjo20** [comment]
Ah you actually beat me to the documentation on this - but I would actually recommend using a custom Authenticator which is the default authenticator on your connector. 

Inside your connector, you should make another request to grab the token and then authenticate the request you were going to send. You’ll also need to build an if-statement to make sure you don’t run into a loop. Here is an example:

```php
<?php

use Saloon\Contracts\Authenticator;
use Saloon\Http\Auth\TokenAuthenticator;

class ForgeConnector extends Connector
{ 
    public function resolveBaseUrl(): string
    {
        return 'https://forge.laravel.com/api/v1';
    }

    protected function defaultAuth(): ?Authenticator
    {
        return new ForgeAuthenticator();
    }
}


// 

class ForgeAuthenticator implements Authenticator
{
    public function set(PendingRequest $pendingRequest): void
    {
				// Ignore if the request is an AuthenticateRequest to prevent loops

				if ($pendingRequest->getRequest() instanceof AuthenticateRequest) {
						return;
				}

				// Make a request to the Authentiation endpoint

				$response = $pendingRequest->getConnector()->send(new AuthenticateRequest);
				
				// Add the token to the request headers

        $pendingRequest->headers()->add('Authorization', 'Bearer ' . $response->json('token'));
    }
}
```

Let me know your thoughts and if this works for you - if it does I’ll add it to the official docs

**Comment by labomatik** [comment, marked as answer]
That's exactly that ;-) it's working correctly, amazing that in few files and lines you handle the complexity to connect to an API...

Many thanks for your package and support!

**Reply by Sammyjo20** [reply]
Amazing ❤️ I’m thinking of building an FAQ on the docs for these small problems. Super glad you like Saloon and thank you so much for the donation!

**Reply by Sammyjo20** [reply]
I have added it to the documentation here: https://docs.saloon.dev/the-basics/authentication#authenticating-apis-that-require-per-request-authentication

You may need to clear cache to view it
